### PR TITLE
Fix 77603 - Don't break the log line when it is more than 16K

### LINKED ISSUE #77603: Pod logs: long lines are corrupted when using timestamps=true
**What happened**:

Docker has a hardcoded 16K limit on log line length, which splits lines longer than that into multiple messages in the log file. `ReadLogs` in [logs.go](https://github.com/kubernetes/kubernetes/blob/255a73ffc64d1c4e0f9a378ebf8d0005edfe6962/pkg/kubelet/kuberuntime/logs/logs.go), however, doesn't account for this when `timestamps` is set to `true` and prints the timestamp in front of every entry in the log file instead of every logical line. This leads to random timestamps appearing in the middle of log lines, potentially breaking the applications that parse these logs.

**What you expected to happen**:

`ReadLogs` inserts timestamps once per logical line, so original log data is preserved. Alternatively, allow `/logs` API to request the logs in a sane format like JSON, which would allow merging the lines back and combining them with the timestamps more intelligently on the application side.

**How to reproduce it (as minimally and precisely as possible)**:

* Create `foo.yaml`:

      apiVersion: v1
      kind: Pod
      metadata:
        name: log-lines
      spec:
        containers:
        - name: main
          image: busybox
          command: ['sh', '-c', 'while sleep 1; do echo -n ___; for i in $(seq 1 18000); do echo -n A; done; echo ___; done']

 * Start the pod with `kubectl apply -f foo.yaml`
 * Fetch the logs with `kubectl logs log-lines --timestamps`, or using the API (`curl https://…/api/v1/namespaces/default/pods/log-lines/log?timestamps=true&container=main`)
 * Observe random timestamps inserted in the middle of the log lines.

**Environment**:
- Kubernetes version (use `kubectl version`):
```Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-19T22:12:47Z", GoVersion:"go1.12.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.5", GitCommit:"2166946f41b36dea2c4626f90a77706f426cdea2", GitTreeState:"clean", BuildDate:"2019-03-25T15:19:22Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
```

### PR DESCRIPTION
**Pull request by YoubingLi**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

When the log line's content size is more than 16K, "kubectl logs --timestamps"  breaks the line as multiple lines .

It is not convenient for other tool to parse the log

**Which issue(s) this PR fixes**:

Fixes #77603 

**Special notes for your reviewer**:
none

```release-note
none
```
/cc @aermakov-zalando

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
thanks for the PR @YoubingLi 
added one suggestion.

/priority important-longterm

File: pkg/kubelet/kuberuntime/logs/logs.go [hunk 1]
```diff
@@ -376,14 +377,19 @@ func ReadLogs(ctx context.Context, path, containerID string, opts *LogOptions, r
...
 				klog.V(2).Infof("Finish parsing log file %q, hit bytes limit %d(bytes)", path, opts.bytes)
 				return nil
 			}
 			klog.Errorf("Failed with err %v when writing log for log file %q: %+v", err, path, msg)
 			return err
 		}
+		if msg.log[len(msg.log) - 1] == eol[0] {
+			isNewLine = true
+		} else {
+			isNewLine = false /* The subsequent content is not a newline */
```
Comment (neolit123) [inline pkg/kubelet/kuberuntime/logs/logs.go]: i'd write this as:
```
isNewLine = msg.log[len(msg.log) - 1] == eol[0]
```
and have a comment on top of the line explaining with a couple of sentences what this line does.

**Review by mattjmcnaughton — COMMENTED** [review]
Thanks for this pr :)

Agree with the changes suggested by @neolit123 :)

I'm also wondering if you would be willing to add a unit test to `pkg/kubelet/kuberuntime/logs/logs_test.go` ensuring we handle this case appropriately? I think it would be good to have that guarantee of this behavior being respected going forward.

Thanks again for your work on this :)

**Comment by YoubingLi** [comment]
@neolit123  @mattjmcnaughton 

      All comments have been addressed and please help review again.

File: pkg/kubelet/kuberuntime/logs/logs.go [hunk 2]
```diff
@@ -376,14 +377,16 @@ func ReadLogs(ctx context.Context, path, containerID string, opts *LogOptions, r
...
-		if err := writer.write(msg); err != nil {
+		if err := writer.write(msg, isNewLine); err != nil {
 			if err == errMaximumWrite {
 				klog.V(2).Infof("Finish parsing log file %q, hit bytes limit %d(bytes)", path, opts.bytes)
 				return nil
 			}
 			klog.Errorf("Failed with err %v when writing log for log file %q: %+v", err, path, msg)
 			return err
 		}
+		/* The subsequent content is the rest part of current line or not*/
```
Comment (neolit123) [inline pkg/kubelet/kuberuntime/logs/logs.go]: please change this to:
`// Check if a new line character has been reached`

**Comment by mattjmcnaughton** [comment]
Going to let @neolit123 take the review here, as I was mostly concerned about adding unit test coverage and you did that - so thank you!

File: pkg/kubelet/kuberuntime/logs/logs.go (hunk 2 above)
Comment (YoubingLi) [reply pkg/kubelet/kuberuntime/logs/logs.go]: The comment has been addressed.

Comment (neolit123) [reply pkg/kubelet/kuberuntime/logs/logs.go]: thanks.
please ping the #sig-node channel on slack if you need approval for the PR.
/ok-to-test

**Comment by YoubingLi** [comment]
@vishh @dchen1107 

    Please help review the change.

**Comment by astefanutti** [comment]
@YoubingLi Do you plan to have the PR rebase and eventually merged? I'm willing to help if needed.

**Comment by astefanutti** [comment]
@YoubingLi thanks a lot for rebasing the PR.

@neolit123 @mattjmcnaughton could you review it?

File: pkg/kubelet/kuberuntime/logs/logs.go [hunk 3]
```diff
@@ -192,6 +192,7 @@ func parseDockerJSONLog(log []byte, msg *logMessage) error {
 	}
 	msg.timestamp = l.Created
 	msg.stream = runtimeapi.LogStreamType(l.Stream)
+
```
Comment (neolit123) [inline pkg/kubelet/kuberuntime/logs/logs.go:195]: no need for this new empty line.

**Comment by neolit123** [comment]
/remove lifecycle-stale

hello, this seems fine, but i have no context if it's desired by the kubelet maintainers.
you could ask in the #sig-node channel in k8s slack.

also we are in code-freeze for 1.19 and this can merge only after code freeze is lifted.

**Comment by astefanutti** [comment]
@kubernetes/sig-node-bugs Can someone review this PR?

**Comment by pacoxu** [comment]
@YoubingLi need rebase 😓
/triage accepted

**Comment by SergeyKanzhelev** [comment]
The change looks good and logical to me. 

My worry is that this fix is heavily depends on how logs are reported and changing this behavior may result in unexpected situations. 

What can address this worry is perhaps introduction of a timeout between writes to decide whether it's a new line. If two lines are written more than 5 minutes (30 seconds?) apart, we can fall back to original behavior.

**Pull request closed without merging**